### PR TITLE
Added support for Software Token MFA.

### PR DESCRIPTION
**Pull request by Taikono-Himazin**
I was looking for an easy way to add Software Token MFA from Python. I think adding functionality to this library is the easiest solution.
I think this will solve what has remained as an issue since the warrant era.
https://github.com/capless/warrant/issues/169
However, the function is still incomplete. We will continue to add more if necessary.

Corrections are welcome. Please point out if the function layout or naming convention is incorrect.
Also, I would like to mention it in the document if necessary, but please tell me how to do it.
thank you.

Since it is Google Translate, there may be something wrong. I'm sorry.

### CODE REVIEW

**Comment by pvizeli** [comment]
Would be great if you could update the readme with an example. thanks

**Review by pvizeli — APPROVED** [review]

File: pycognito/__init__.py [hunk 1]
```diff
@@ -889,3 +895,106 @@ def admin_update_user_pool_client(self, pool_id: str, client_id: str, **kwargs):
...
+        """
+        sms_mfa_settings = {"Enabled": bool(sms_mfa), "PreferredMfa": False}
+        software_token_mfa_settings = {
+            "Enabled": bool(software_token_mfa),
+            "PreferredMfa": False,
+        }
+        if not (bool(sms_mfa) or bool(software_token_mfa)):
+            # Disable MFA
+            pass
+        if preferred == "SMS":
```
Comment (nk9) [inline pycognito/__init__.py:940]: The parameter doc says "However, `preferred` is not needed only if both of the previous arguments are False." So I think this supposed to be `elif`?

File: pycognito/__init__.py [hunk 2]
```diff
@@ -889,3 +895,106 @@ def admin_update_user_pool_client(self, pool_id: str, client_id: str, **kwargs):
...
+        :rtype: string
+        """
+        response = self.client.associate_software_token(AccessToken=self.access_token)
+        return response["SecretCode"]
+
+    def verify_software_token(self, code, device_name):
+        """
+        Verify the value generated by TOTP to complete the registration of Software Token MFA.
+        :param code: The value generated by TOTP
+        :param device_name: Device name to register (optional)
```
Comment (nk9) [inline pycognito/__init__.py:912]: The parameter is declared as optional, but doesn't have a default value.

File: pycognito/__init__.py [hunk 3]
```diff
@@ -889,3 +895,106 @@ def admin_update_user_pool_client(self, pool_id: str, client_id: str, **kwargs):
...
+            AccessToken=self.access_token,
+        )
+
+    def respond_to_software_token_mfa_challenge(self, code, mfa_tokens=None):
+        """
+        Respons challenge to software token of MFA.
+        :param code: software token MFA code.
+        :type code: string
+        :param code: mfa_token stored in MFAChallengeException. Not required if you have not regenerated the Cognito instance.
+        :type code: string
```
Comment (nk9) [inline pycognito/__init__.py:960]: The parameter name is `mfa_tokens`, but the doc calls it `code`. And it's some collection, not a string, right?

File: pycognito/__init__.py [hunk 4]
```diff
@@ -889,3 +895,106 @@ def admin_update_user_pool_client(self, pool_id: str, client_id: str, **kwargs):
...
+        )
+        self._set_tokens(tokens)
+
+    def respond_to_sms_mfa_challenge(self, code, mfa_tokens=None):
+        """
+        Respons challenge to SMS MFA.
+        :param code: SMS MFA code.
+        :type code: string
+        :param code: mfa_token stored in MFAChallengeException. Not required if you have not regenerated the Cognito instance.
+        :type code: string
```
Comment (nk9) [inline pycognito/__init__.py:984]: Same as above.

File: pycognito/__init__.py [hunk 5]
```diff
@@ -889,3 +895,106 @@ def admin_update_user_pool_client(self, pool_id: str, client_id: str, **kwargs):
...
+            ClientId=self.client_id,
+            Session=mfa_tokens["Session"],
+            ChallengeName="SOFTWARE_TOKEN_MFA",
+            ChallengeResponses=challenge_responses,
+        )
+        self._set_tokens(tokens)
+
+    def respond_to_sms_mfa_challenge(self, code, mfa_tokens=None):
+        """
+        Respons challenge to SMS MFA.
```
Comment (nk9) [inline pycognito/__init__.py:980]: I would say "Sends a response to an SMS MFA challenge."

**Review by nk9 — COMMENTED** [review]
Sorry for the late review, but I think I've found a logic issue.

**Pull request merged**